### ISSUE TITLE
Is it possible to use the current implementation with Futures 0.3?

### ISSUE THREAD

**Issue by fgimian** [eclipse/paho.mqtt.rust#66]
Hey there, thanks so much for your work on this library 😄 

I understand that Futures 0.3 support is coming later in the year, but was just curious if you have an example of using this library with Futures 0.3?

My scenario is that I would like to take each incoming message, process it and insert into a database.   Of course, the main point is to avoid blocking further messages from being received.

I'm getting lots of errors about certain traits not being implemented when running against Futures 0.3 (even though I have `use`d them).  These are for traits such as `and_then` and `for_each`.

I'm still pretty new to Rust and am a bit lost if this is even possible.  Any help would be greatly appreciated?

Thanks heaps
Fotis

**Comment by fgimian** [comment]
Alrgihty, I've spent the day on this and got it mostly working with one caveat.

My **Cargo.toml** dependencies look like this:

```ini
[dependencies]
chrono = "0.4.10"
futures = { version = "0.3.4", features = ["compat"] }
paho-mqtt = { git = "https://github.com/eclipse/paho.mqtt.rust.git" }
tokio = { version =  "0.2.11", features = ["full"] }
```

And my **main.rs** is as follows:

```rust
#![warn(rust_2018_idioms)]

use std::time;

use futures::{
    compat::{Future01CompatExt, Stream01CompatExt},
    stream::StreamExt,
};
use paho_mqtt as mqtt;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let host = "ssl://localhost".to_string();
    println!("Connecting to host: '{}'", host);

    // Configure the connection options (including SSL and LWT).
    let ssl_options = mqtt::SslOptionsBuilder::new()
        .trust_store("certs/cacert.pem")
        .key_store("certs/client-cert.pem")
        .private_key("certs/client-key.pem")
        .finalize();

    let lwt = mqtt::Message::new("failure", "Futures consumer lost connection", 1);
    let connect_options = mqtt::ConnectOptionsBuilder::new()
        .ssl_options(ssl_options)
        .mqtt_version(mqtt::MQTT_VERSION_3_1_1)
        .keep_alive_interval(time::Duration::from_secs(30))
        // Automation reconnect takes two arguments: min_retry_interval and max_retry_interval.
        .automatic_reconnect(
            time::Duration::from_millis(500),
            time::Duration::from_secs(30),
        )
        // Use a persistent session so we can recieve missed messages when reconnecting.
        .clean_session(false)
        .will_message(lwt)
        .finalize();

    // Create our MQTT client.
    let mut client = mqtt::AsyncClientBuilder::new()
        .server_uri(&host)
        .client_id("mqtt-client-rust")
        .finalize();

    // Open the receiver stream before connecting to ensure we don't miss any messages, even when
    // a restart of the app gets us a flood of stored messages from the persistent session.
    let mut rx = client.get_stream(10).compat();

    let connect_response = client.connect(connect_options).compat().await?;
    if let Some((server_uri, version, session_present)) = connect_response.connect_response() {
        println!("Connection response received");
        println!("Server URI: {}", server_uri);
        println!("Version: {}", version);
        println!("Session Present: {}", session_present);

        // Subscribe to appropriate topics if we don't have an existing persistent session.
        if !session_present {
            let topics = ["data/#"];
            let qos = [2];
            let subscribe_response = client.subscribe_many(&topics, &qos).compat().await?;
            if let Some(qos) = subscribe_response.subscribe_many_response() {
                if !qos.is_empty() {
                    println!("QoS granted for subscribed topics: {:?}", qos);
                }
            }
        }
    }

    while let Some(Ok(optional_message)) = rx.next().await {
        tokio::spawn(async move {
            match optional_message {
                Some(message) => {
                    let now = chrono::Local::now().format("%Y-%m-%d %H:%M:%S");
                    println!("{} Received message on topic: {:?}", now, message.topic());
                    println!("{} Received message: {:?}", now, message.payload());
                }
                // TODO: Should we handle these cases differently?
                None => {
                    println!("Stream disruption");
                }
            }
        });
    }

    Ok(())
}
```

Now this works perfectly except for one important aspect.  If I interrupt the connection, reconnection is triggered but this causes as Segmentation fault.  I'm really not sure why this happens and I can confirm it definitely doesn't using the [example in this repo](https://github.com/eclipse/paho.mqtt.rust/blob/master/examples/futures_consume.rs) with Futures 0.1.

As a workaround, I disabled automatic_reconnect in my connect options and instead handled it myself as follows:

```rust
    tokio::spawn(async move {
        while let Some(Ok(optional_message)) = rx.next().await {
            tokio::spawn(async move {
                match optional_message {
                    Some(message) => {
                        let now = chrono::Local::now().format("%Y-%m-%d %H:%M:%S");
                        println!("{} Received message on topic: {:?}", now, message.topic());
                        println!("{} Received message: {:?}", now, message.payload());
                    }
                    // TODO: Should we handle these cases differently?
                    None => {
                        println!("Stream disruption");
                    }
                }
            });
        }
    });

    loop {
        if !client.is_connected() {
            println!("Lost connection, reconnecting!");
            client.reconnect().compat().await.ok();
        }
        std::thread::sleep(std::time::Duration::from_secs(1));
    }
```

What do you think?  Any idea why the first example causes a Segmentation fault?

Huge thanks in advance! 😄 
Fotis

**Comment by fgimian** [comment]
I've attempted to capture more details using gdb in case this helps:

```
Thread 7 "mqtt-mon" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff583d700 (LWP 17147)]
0x000055555568aa40 in core::task::wake::Waker::wake_by_ref (self=0x555555a24260)
    at /rustc/433aae93e4ef866a1fdfefad136b32ed89acd3e7/src/libcore/task/wake.rs:256
256     /rustc/433aae93e4ef866a1fdfefad136b32ed89acd3e7/src/libcore/task/wake.rs: No such file or directory.
(gdb)
Continuing.

Thread 7 "mqtt-mon" received signal SIGSEGV, Segmentation fault.
0x000055555568aa40 in core::task::wake::Waker::wake_by_ref (self=0x555555a24260)
    at /rustc/433aae93e4ef866a1fdfefad136b32ed89acd3e7/src/libcore/task/wake.rs:256
256     in /rustc/433aae93e4ef866a1fdfefad136b32ed89acd3e7/src/libcore/task/wake.rs
(gdb)
Continuing.
[Thread 0x7ffff583d700 (LWP 17147) exited]
[Thread 0x7ffff603e700 (LWP 17146) exited]
[Thread 0x7ffff623f700 (LWP 17145) exited]
[Thread 0x7ffff6440700 (LWP 17144) exited]
[Thread 0x7ffff6641700 (LWP 17143) exited]
[Thread 0x7ffff6842700 (LWP 17142) exited]

Program terminated with signal SIGSEGV, Segmentation fault.
The program no longer exists.
```

Thanks a lot
Fotis

**Comment by fpagliughi** [comment]
I haven't tried any of this. I just assumed the compat layer would work, but really, I'm anxious to get the library up to 0.3 async/await support. There is not a lot of work required in the library itself, but I want to get the MQTT v5 support out the door, and unfortunately, even though it's mostly done, a release is being delayed by some upstream work in the C lib. Hopefully that will be done soon.

Meanwhile I will take a look at this at my earliest convenience. It does look like there may be some missing thread synchronization in the futures completion code.

**Comment by fgimian** [comment]
Thank you so much for the reply @fpagliughi.  I'm happy to work around this for now as it's not really the end game anyway.  As soon as Futures 0.3 support comes along in this library, I'll simply drop the compat layer and the workaround.  So it may not be worth spending too much of your time on tracking down this issue when using the compat layer.

I ended up with this implementation (instead of the one above) which I think is sufficiently elegant:

```rust
    // Monitor the connection every second and reconnect if required.
    tokio::spawn(async move {
        loop {
            if !client.is_connected() {
                println!("Client is not connected, attempting reconnect");
                match client.reconnect().compat().await {
                    Ok(result) => println!("Successfully reconnected: {:?}", result),
                    Err(error) => println!("Failed to reconnect: {:?}", error),
                }
            }
            tokio::time::delay_for(time::Duration::from_secs(1)).await;
        }
    });

    // Consume messages and process them in a future.
    while let Some(Ok(optional_message)) = rx.next().await {
        tokio::spawn(async move {
            match optional_message {
                Some(message) => {
                    let now = chrono::Local::now().format("%Y-%m-%d %H:%M:%S");
                    println!("{} Received message on topic: {:?}", now, message.topic());
                    println!("{} Received message: {:?}", now, message.payload());
                }
                // TODO: Should we handle these cases differently?
                None => {
                    println!("Stream disruption");
                }
            }
        });
    }
```

Huge thanks for your reply and help! 😄 
Fotis

**Comment by fpagliughi** [comment]
Sure. But let's keep the issue open for now. I will want to look at it more closely soon to make sure that there's not a bug that would carry over to the 0.3 implementation.

**Comment by fgimian** [comment]
> Sure. But let's keep the issue open for now. I will want to look at it more closely soon to make sure that there's not a bug that would carry over to the 0.3 implementation.

Absolutely, no worries at all 😄 

Thanks again!
Fotis

**Comment by FanOfSilence** [comment]
> Alrgihty, I've spent the day on this and got it mostly working with one caveat.
> 
> My **Cargo.toml** dependencies look like this:
> 
> ```ini
> [dependencies]
> chrono = "0.4.10"
> futures = { version = "0.3.4", features = ["compat"] }
> paho-mqtt = { git = "https://github.com/eclipse/paho.mqtt.rust.git" }
> tokio = { version =  "0.2.11", features = ["full"] }
> ```
> 
> And my **main.rs** is as follows:
> 
> ```rust
> #![warn(rust_2018_idioms)]
> 
> use std::time;
> 
> use futures::{
>     compat::{Future01CompatExt, Stream01CompatExt},
>     stream::StreamExt,
> };
> use paho_mqtt as mqtt;
> 
> #[tokio::main]
> async fn main() -> Result<(), Box<dyn std::error::Error>> {
>     let host = "ssl://localhost:8883".to_string();
>     println!("Connecting to host: '{}'", host);
> 
>     // Configure the connection options (including SSL and LWT).
>     let ssl_options = mqtt::SslOptionsBuilder::new()
>         .trust_store("certs/cacert.pem")
>         .key_store("certs/client-cert.pem")
>         .private_key("certs/client-key.pem")
>         .finalize();
> 
>     let lwt = mqtt::Message::new("failure", "Futures consumer lost connection", 1);
>     let connect_options = mqtt::ConnectOptionsBuilder::new()
>         .ssl_options(ssl_options)
>         .mqtt_version(mqtt::MQTT_VERSION_3_1_1)
>         .keep_alive_interval(time::Duration::from_secs(30))
>         // Automation reconnect takes two arguments: min_retry_interval and max_retry_interval.
>         .automatic_reconnect(
>             time::Duration::from_millis(500),
>             time::Duration::from_secs(30),
>         )
>         // Use a persistent session so we can receive missed messages when reconnecting.
>         .clean_session(false)
>         .will_message(lwt)
>         .finalize();
> 
>     // Create our MQTT client.
>     let mut client = mqtt::AsyncClientBuilder::new()
>         .server_uri(&host)
>         .client_id("mqtt-client-rust")
>         .finalize();
> 
>     // Open the receiver stream before connecting to ensure we don't miss any messages, even when
>     // a restart of the app gets us a flood of stored messages from the persistent session.
>     let mut rx = client.get_stream(10).compat();
> 
>     let connect_response = client.connect(connect_options).compat().await?;
>     if let Some((server_uri, version, session_present)) = connect_response.connect_response() {
>         println!("Connection response received");
>         println!("Server URI: {}", server_uri);
>         println!("Version: {}", version);
>         println!("Session Present: {}", session_present);
> 
>         // Subscribe to appropriate topics if we don't have an existing persistent session.
>         if !session_present {
>             let topics = ["data/#"];
>             let qos = [2];
>             let subscribe_response = client.subscribe_many(&topics, &qos).compat().await?;
>             if let Some(qos) = subscribe_response.subscribe_many_response() {
>                 if !qos.is_empty() {
>                     println!("QoS granted for subscribed topics: {:?}", qos);
>                 }
>             }
>         }
>     }
> 
>     while let Some(Ok(optional_message)) = rx.next().await {
>         tokio::spawn(async move {
>             match optional_message {
>                 Some(message) => {
>                     let now = chrono::Local::now().format("%Y-%m-%d %H:%M:%S");
>                     println!("{} Received message on topic: {:?}", now, message.topic());
>                     println!("{} Received message: {:?}", now, message.payload());
>                 }
>                 // TODO: Should we handle these cases differently?
>                 None => {
>                     println!("Stream disruption");
>                 }
>             }
>         });
>     }
> 
>     Ok(())
> }
> ```
> 
> Now this works perfectly except for one important aspect. If I interrupt the connection, reconnection is triggered but this causes a Segmentation fault. This only happens when reconnection fails. If I bring my server back online quickly, the client re-establishes a connection and doesn't segfault.
> 
> I'm really not sure why this happens and I can confirm it definitely doesn't using the [example in this repo](https://github.com/eclipse/paho.mqtt.rust/blob/master/examples/futures_consume.rs) with Futures 0.1.
> 
> As a workaround, I disabled automatic_reconnect in my connect options and instead handled it myself as follows:
> 
> ```rust
>     tokio::spawn(async move {
>         while let Some(Ok(optional_message)) = rx.next().await {
>             tokio::spawn(async move {
>                 match optional_message {
>                     Some(message) => {
>                         let now = chrono::Local::now().format("%Y-%m-%d %H:%M:%S");
>                         println!("{} Received message on topic: {:?}", now, message.topic());
>                         println!("{} Received message: {:?}", now, message.payload());
>                     }
>                     // TODO: Should we handle these cases differently?
>                     None => {
>                         println!("Stream disruption");
>                     }
>                 }
>             });
>         }
>     });
> 
>     loop {
>         if !client.is_connected() {
>             println!("Lost connection, reconnecting!");
>             client.reconnect().compat().await.ok();
>         }
>         std::thread::sleep(std::time::Duration::from_secs(1));
>     }
> ```
> 
> What do you think? Any idea why the first example causes a Segmentation fault?
> 
> Huge thanks in advance! smile
> Fotis

Hi, I'm trying to do the same thing right now. I tried this example and it compiles fine. But did you ever figure out what type to put on ```rx``` to get it to compile if you are receiving it as an argument in a function?
I can't get that to work because ```Receiver``` has been moved from ```futures::sync::mpsc``` in version 0.1 to ```futures::channel::mpsc``` in version 0.3. ```compat``` doesn't seem to solve that issue.

**Comment by fpagliughi** [comment]
I've been waiting on an upstream release of the Paho C library which just went out a few days ago. With that done, I'm going to push out basically what has been sitting in the `master` branch for the last few months as v0.7 - basically have an official release for MQTT v5. This will still be Futures 0.1.

But then I'm going to immediately start work on the next version with async/await (std futures) support.

**Comment by fgimian** [comment]
Thank you so much @fpagliughi 😄

**Comment by fpagliughi** [comment]
It'll be messy for the next few weeks, but I'm pushing some updates on this into the `develop` branch now.

**Comment by fpagliughi** [comment]
@ryankurte : Apologies, I totally forgot about this, as it was buried at the end of this Issue that I already marked "fixed":

> just a heads up that the abstract return type here makes embedding the client in higher-level objects a bit of a pain, it might be easier to just return the mpsc::Receiver as with the prior version (or to implement Stream on the Client object, but, same difference really)

I opened a new issue for it, #107.
I'll consider reverting it back in the next rev which should be out in a few weeks.
